c++/src/Reader.cc

```diff
@@ -944,13 +1035,67 @@ namespace orc {
     // update row number
     previousRow = firstRowOfStripe[currentStripe] + currentRowInStripe;
     currentRowInStripe += rowsToRead;
+
+    // check if we need to advance to next selected row group
+    if (sargsApplier) {
+      uint64_t nextRowToRead = advanceToNextRowGroup(currentRowInStripe,
+                                                     rowsInCurrentStripe,
+                                                     footer->rowindexstride(),
+                                                     sargsApplier->getRowGroups());
+      if (currentRowInStripe != nextRowToRead) {
+        // it is guaranteed to be at start of a row group
+        currentRowInStripe = nextRowToRead;
+        if (currentRowInStripe < rowsInCurrentStripe) {
+          seekToRowGroup(static_cast<uint32_t>(currentRowInStripe / footer->rowindexstride()));
```

Comment:
I think that there is a hidden performance regression here:
at the end of the call chain, e.g. ColumnReader::seekToRowGroup(), RleDecoderV2::seek(), ZlibDecompressionStream::seek(), the inputBuffer is reset to nullptr:
https://github.com/apache/orc/blob/028261ad6e26d8ae9f9483c8aeb76a85c7f1168b/c%2B%2B/src/Compression.cc#L573
This leads to calling ZlibDecompressionStream::readBuffer() and reading from the stream again, even if the whole stream was buffered previously to inputBuffer (which is usually the case in Impala, as we set the block size to 8 MB, but can also cause problems with the default 256KB).
In the worst case this can lead to reading the whole column O(num_of_row_groups) times. 

The solution could be to make ZlibDecompressionStream/BlockDecompressionStream::seek() smarter and keep the inputBuffer if the new position (which is the header byte of the new compression block) is still in it.

A further improvement could be to make ZlibDecompressionStream/BlockDecompressionStream::Skip() smarter to avoid decompressing all the skipped blocks in this case ( https://github.com/apache/orc/blob/028261ad6e26d8ae9f9483c8aeb76a85c7f1168b/c%2B%2B/src/Compression.cc#L545 ), but this is not a regression so can go to another patch.

Reply:
You are right. Performance of seeking is another headache if you use large compression blocks. In our internal ORC setup, we have optmized this by finishing compression blocks at the end of every row group. In this way, seeking will never read compression blocks of other row groups.

Reply:
I have created a Jira (ORC-614) for the optimization of seek(), and we put together an implementation with Norbert, he will upload it soon as pull request. It seems to solve the regressions we experienced.

About finishing compression blocks at row group end: I actually created a ticket (IMPALA-8449) to do something similar in Parquet:
In Parquet we came from the other direction - pages are both the unit of indexing and compression in Parquet, but they are not aligned, so unlike ORC, index entries can point to different row numbers for every column. This decreases decreases the efficiency of combining filters for different columns + makes the reader implementation very complex.